crates/mev-share-rpc-api/src/auth.rs

```diff
@@ -0,0 +1,199 @@
+//! A layer responsible for implementing flashbots-style authentication
+//! by signing the request body with a private key and adding the signature
+//! to the request headers.
+
+use std::{
+    error::Error,
+    task::{Context, Poll},
+};
+
+use ethers_core::{types::H256, utils::keccak256};
+use ethers_signers::Signer;
+use futures_util::future::BoxFuture;
+
+use http::{header::HeaderValue, HeaderName, Request};
+use hyper::Body;
+
+use tower::{Layer, Service};
+
+const FLASHBOTS_HEADER: HeaderName = HeaderName::from_static("x-flashbots-signature");
+
+/// Layer that applies [`FlashbotsSigner`] which adds a request header with a signed payload.
+#[derive(Clone)]
+pub struct FlashbotsSignerLayer<S> {
+    signer: S,
+}
+
+impl<S> FlashbotsSignerLayer<S> {
+    /// Creates a new [`FlashbotsSignerLayer`] with the given signer.
+    pub fn new(signer: S) -> Self {
+        FlashbotsSignerLayer { signer }
+    }
+}
+
+impl<S: Clone, I> Layer<I> for FlashbotsSignerLayer<S> {
+    type Service = FlashbotsSigner<S, I>;
+
+    fn layer(&self, inner: I) -> Self::Service {
+        FlashbotsSigner { signer: self.signer.clone(), inner }
+    }
+}
+
+/// Middleware that adds a request header with a signed payload.
+/// For more info, see https://docs.flashbots.net/flashbots-auction/searchers/advanced/rpc-endpoint#authentication
```

Comment:
this should mention that this will sign the request body and set the flashbots header